source/common/http/utility.cc

```diff
@@ -562,5 +565,42 @@ void Utility::traversePerFilterConfigGeneric(
   }
 }
 
+std::string Utility::PercentEncoding::encode(absl::string_view value) {
+  std::ostringstream encoded;
+  for (size_t i = 0; i < value.size(); ++i) {
+    char c = value[i];
+    // Unreserved characters. unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~".
+    // https://tools.ietf.org/html/rfc3986#section-2.3.
+    if (absl::ascii_isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
+      encoded << c;
+      // TODO(dio): Check for ' ' and allow to encode it as '+' instead of '%20' if desired.
+    } else {
+      // For consistency, URI producers should use uppercase hexadecimal digits for all
+      // percent-encodings. https://tools.ietf.org/html/rfc3986#section-2.1.
+      encoded << std::uppercase;
+      encoded << '%' << std::hex << std::setw(2) << std::setfill('0') << (c & 0xff) << std::dec;
+      encoded << std::nouppercase;
+    }
+  }
+  return encoded.str();
+}
+
+std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
+  std::ostringstream decoded;
+  for (size_t i = 0; i < encoded.size(); ++i) {
+    char c = encoded[i];
+    if (c == '%') {
+      int d;
+      std::istringstream captured(std::string(encoded.substr(i + 1, 2)));
```

Comment:
Will `substr` blow up here in case the input is malformed (ie there is a % with less than 2 trailing characters)?